sequencing-server/src/lib/codegen/CommandEDSLPreface.ts

```diff
@@ -3233,6 +3233,36 @@ export function doyToInstant(doy: DOY_STRING): Temporal.Instant {
 export function durationToHms(time: Temporal.Duration): HMS_STRING {
   let { days, hours, minutes, seconds, milliseconds } = time;
 
+  while (milliseconds >= 1000) {
+    seconds += 1;
+    milliseconds -= 1000;
+  }
+
+  while (milliseconds <= -1000) {
+    seconds -= 1;
+    milliseconds += 1000;
+  }
+
+  while (seconds >= 60) {
+    minutes += 1;
+    seconds -= 60;
+  }
+
+  while (seconds <= -60) {
+    minutes -= 1;
+    seconds += 60;
+  }
+
+  while (minutes >= 60) {
+    hours += 1;
+    minutes -= 60;
+  }
+
+  while (minutes <= -60) {
+    hours -= 1;
+    minutes += 60;
+  }
```

Comment:
```suggestion
function normalize(lg_time: number, sm_time: number, ratio: number): [number, number] {
    let sm = sm_time % ratio;
    return [lg_time + (sm_time - sm) / ratio, sm];
}

[seconds, milliseconds] = normalize(seconds, milliseconds, 1000)
[minutes, seconds] = normalize(minutes, seconds, 60)
[hours, minutes] = normalize(hours, minutes, 60)
```
etc. will do the same thing without a loop.

Just to double check the logic for negative times, though: This leaves negative time components negative, just clamped to within the appropriate unit. Is that what we want?

Reply:
@mattdailis This might have already been merged!